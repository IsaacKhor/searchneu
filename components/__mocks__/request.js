/*
 * This file is part of Search NEU and licensed under AGPL3.
 * See the license file in the root folder for details.
 */

import URI from "urijs";
import macros from "../macros";

// This file mocks out request.js
// Provides stubs for all the methods.

const searchExample = {
  results: [
    {
      score: 4.009930822280515,
      type: "class",
      class: {
        crns: [],
        classAttributes: ["GS Col of Arts", "Media & Design"],
        maxCredits: 4,
        minCredits: 4,
        desc:
          "Examines time-tested and cutting-edge methods for shaping and presenting messages across multimedia platforms to effectively disseminate an organization’s message, change a public conversation, or shift public opinion. Examines case studies in mainstream media, public advocacy, and strategic communications to explore the motivations and methods of the organizations as well as the tools and techniques used. Examines the practice of digital advocacy by exploring and applying pertinent findings from politics, advertising, and behavioral science that are increasingly employed by professionals looking to “micro-target” voters, “convert” customers, or “nudge” the public. One major component of the course is hands-on workshops through which students are offered an opportunity to learn how to leverage the latest digital tools for communicating across social media and online platforms. 4.000 Lecture hours",
        classId: "5400",
        prettyUrl:
          "https://wl11gp.neu.edu/udcprod8/bwckctlg.p_disp_course_detail?cat_term_in=202010&subj_code_in=JRNL&crse_numb_in=5400",
        name: "Media and Advocacy in Theory and Practice",
        url:
          "https://wl11gp.neu.edu/udcprod8/bwckctlg.p_disp_listcrse?term_in=202010&subj_in=JRNL&crse_in=5400&schd_in=%",
        lastUpdateTime: 1565626142874,
        termId: "202010",
        host: "neu.edu",
        subject: "JRNL",
        optPrereqsFor: {
          values: [],
        },
        prereqsFor: {
          values: [],
        },
      },
      sections: [],
    },
    {
      score: 3.2079446578244117,
      type: "class",
      class: {
        crns: [],
        classAttributes: ["NUpath Analyzing/Using Data", "Engineering"],
        prereqs: {},
        coreqs: {},
        maxCredits: 4,
        minCredits: 4,
        desc:
          "Introduces basic measurements and data analysis techniques. Offers students an opportunity to become familiar with various types of measurement systems and to set up and perform experiments according to a given procedure. Covers basic measurement methods of rotational frequency; temperature, pressure, and power; and analog-to-digital conversion techniques and data acquisition. Data analysis topics include statistical analysis of data, probability and inherent uncertainty, basic measurement techniques, primary and secondary standards, system response characteristics, and computerized data acquisition methods. Includes experiments in thermodynamics, fluid mechanics, and heat transfer. Topics include cycle performance, flow discharge coefficient and heat transfer coefficient measurements, and psychometric applications in the air-conditioning field. 4.000 Lecture hours",
        classId: "4505",
        prettyUrl:
          "https://wl11gp.neu.edu/udcprod8/bwckctlg.p_disp_course_detail?cat_term_in=202010&subj_code_in=ME&crse_numb_in=4505",
        name: "Measurement and Analysis with Thermal Science Application",
        url:
          "https://wl11gp.neu.edu/udcprod8/bwckctlg.p_disp_listcrse?term_in=202010&subj_in=ME&crse_in=4505&schd_in=%",
        lastUpdateTime: 1565626142878,
        termId: "202010",
        host: "neu.edu",
        subject: "ME",
        optPrereqsFor: {
          values: [],
        },
        prereqsFor: {
          values: [],
        },
      },
      sections: [],
    },
    {
      score: 2.1209753532061653,
      type: "class",
      class: {
        crns: [],
        classAttributes: ["GSEN Engineering"],
        desc:
          "Introduces the relatively young field of soft matter, which encompasses the physical description of various states of soft materials including liquids, colloids, polymers, foams, gels, granular materials, and a number of biological materials. Soft matter (also known as “soft condensed matter” or “complex fluids”) is less ordered than metals and oxides (hard condensed matter) and is more subject to thermal fluctuations and applied forces. Focuses on critical thinking, problem diagnosis, estimation, statistical analysis, and data-based decision making. Includes many in-class demonstrations from colloidal assembly to emulsion stability to cellular apoptosis. Highlights applications such as industrial processing, life sciences, and environmental remediation. Requires graduate study in related field or permission of instructor. 4.000 Lecture hours",
        classId: "6250",
        prettyUrl:
          "https://wl11gp.neu.edu/udcprod8/bwckctlg.p_disp_course_detail?cat_term_in=202010&subj_code_in=MATL&crse_numb_in=6250",
        name: "Soft Matter",
        url:
          "https://wl11gp.neu.edu/udcprod8/bwckctlg.p_disp_listcrse?term_in=202010&subj_in=MATL&crse_in=6250&schd_in=%",
        lastUpdateTime: 1565626115527,
        maxCredits: 4,
        minCredits: 4,
        termId: "202010",
        host: "neu.edu",
        subject: "MATL",
        optPrereqsFor: {
          values: [],
        },
        prereqsFor: {
          values: [],
        },
      },
      sections: [],
    },
    {
      score: 1.9514106966879916,
      type: "class",
      class: {
        crns: [],
        classAttributes: ["GSEN Engineering"],
        maxCredits: 4,
        minCredits: 4,
        desc:
          "Covers fundamentals of materials thermodynamics that encompass the first, second, and third laws, entropy, enthalpy, and free energy. Emphasis is on phase stability and equilibria, phase diagram computation with applications to phases in metals, alloys, and ionic compounds. Requires knowledge of thermodynamics course and materials science course. 4.000 Lecture hours",
        classId: "7355",
        prettyUrl:
          "https://wl11gp.neu.edu/udcprod8/bwckctlg.p_disp_course_detail?cat_term_in=202010&subj_code_in=MATL&crse_numb_in=7355",
        name: "Thermodynamics of Materials",
        url:
          "https://wl11gp.neu.edu/udcprod8/bwckctlg.p_disp_listcrse?term_in=202010&subj_in=MATL&crse_in=7355&schd_in=%",
        lastUpdateTime: 1565626142876,
        termId: "202010",
        host: "neu.edu",
        subject: "MATL",
        optPrereqsFor: {
          values: [],
        },
        prereqsFor: {
          values: [
            {
              subject: "MATL",
              classId: "7360",
            },
          ],
        },
      },
      sections: [],
    },
    {
      score: 1.88289045341538,
      type: "class",
      class: {
        crns: [],
        classAttributes: ["GSEN Engineering"],
        prereqs: {},
        maxCredits: 4,
        minCredits: 4,
        desc:
          "Covers the processing of metallic and ceramic materials from particulate form. Includes particulate fabrication, characterization, handling, and consolidation for alloys, ceramics, and composites. Other topics include the principles of sintering in the absence and presence of liquid, advanced materials processing by rapid-solidification powder metallurgy, and the processing and structures of advanced ceramics. 4.000 Lecture hours",
        classId: "5380",
        prettyUrl:
          "https://wl11gp.neu.edu/udcprod8/bwckctlg.p_disp_course_detail?cat_term_in=202010&subj_code_in=MATL&crse_numb_in=5380",
        name: "Particulate Materials Processing",
        url:
          "https://wl11gp.neu.edu/udcprod8/bwckctlg.p_disp_listcrse?term_in=202010&subj_in=MATL&crse_in=5380&schd_in=%",
        lastUpdateTime: 1565626142876,
        termId: "202010",
        host: "neu.edu",
        subject: "MATL",
        optPrereqsFor: {
          values: [],
        },
        prereqsFor: {
          values: [],
        },
      },
      sections: [],
    },
  ],
  wasSubjectMatch: false,
};

class MockRequest {
  constructor() {
    // Keeps track of whether to reply to searches for ben
    this.replyWithDataToBen = null;

    this.reset();
  }

  // Reset back to defaults
  reset() {
    this.replyWithDataToBen = true;
  }

  // Sets whether this mock should return the data or return an empty array when
  // 'ben' is searched for.
  // Only effects searches for 'ben'
  setBenResponse(bool) {
    this.replyWithDataToBen = bool;
  }

  async get(request) {
    const urlParsed = new URI(request);

    const parsedQuery = urlParsed.query(true);

    if (
      urlParsed.path() === "/search" &&
      parsedQuery.query === "ben" &&
      !this.replyWithDataToBen
    ) {
      return {
        results: [],
        wasSubjectMatch: false,
      };
    }

    if (
      !macros.isNumeric(parsedQuery.minIndex) ||
      !macros.isNumeric(parsedQuery.maxIndex)
    ) {
      macros.critical("invalid min index or max index");
    }

    const results = { ...searchExample };

    results.results = searchExample.results.slice(
      parsedQuery.minIndex,
      parsedQuery.maxIndex
    );

    return results;
  }
}

export default new MockRequest();
