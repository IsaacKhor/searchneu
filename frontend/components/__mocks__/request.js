/*
 * This file is part of Search NEU and licensed under AGPL3.
 * See the license file in the root folder for details.
 */

import macros from '../macros';

// This file mocks out request.js
// Provides stubs for all the methods.



class MockRequest {
  
  async get() {
    return {
      results: []
    }
  }



}


export default new MockRequest();
