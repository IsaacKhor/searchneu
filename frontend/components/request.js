import idb from 'idb-keyval';


class Request {


  async getFromCache(url) {

    var start = new Date().getTime();

    let existingValue = await idb.get(url)

    var end = new Date().getTime();

    if (existingValue) {
      console.log('Lookup for',url,'found ',existingValue.length,'in', (end - start),'ms');      
    }
    else {
      console.log('Lookup for',url,'not found in', (end - start),'ms');       
    }

    return existingValue;
  }


  async getFromInternet(url) {
    return new Promise((resolve, reject) => {
      const xmlhttp = new XMLHttpRequest();
      xmlhttp.onreadystatechange = function onreadystatechange() {
        if (xmlhttp.readyState !== 4) {
          return;
        }
        if (xmlhttp.status !== 200) {
          let err;
          if (xmlhttp.statusText) {
            err = xmlhttp.statusText;
          } else if (xmlhttp.response) {
            err = xmlhttp.response;
          } else {
            err = `unknown ajax error${String(xmlhttp.status)}`;
          }

          err += `config = ${JSON.stringify(url)}`;

          console.warn('error, bad code recievied', xmlhttp.status, err, url);

          reject(err);
          return;
        }

        const response = JSON.parse(xmlhttp.response);

        if (response.error) {
          console.warn('ERROR networking error bad reqeust?', url);
        }

        resolve(response);
      };


      xmlhttp.open('GET', url, true);
      xmlhttp.send();
    });
  }

  async waitForPromiseThenSave (internetPromise, url) {
    return;

    // let internetValue = await internetPromise.value


    var start = new Date().getTime();

    idb.set(url, ((await internetPromise).value))

    var end = new Date().getTime();

    console.log('Saving ',url,'took', (end-start),'ms')

  }

  async get(config) {
    if (typeof config === 'string') {
      config = {
        url: config,
      };
    }
    else if (Object.keys(config).length > 2) {
      console.error('Nothing is supported except JSON GET requests with an option for caching in idb.', config);
    }


    if (!config.useCache) {
      return this.getFromInternet(config.url)
    }


    let internetPromise = this.getFromInternet(config.url).then(function(value) {
      return {
        type: 'internet',
        value: value
      }
    })

    let idbPromise = this.getFromCache(config.url).then(function(value) {
      return {
        type: 'idb',
        value: value
      }
    })


    let firstResult = await Promise.race([internetPromise, idbPromise])

    // The expected result
    if (firstResult.type === 'idb') {
      if (firstResult.value) {

        // Wait for the internet value and store it in the cache
        // this.waitForPromiseThenSave(internetPromise, config.url);

        console.log('returning', firstResult.value)
        return firstResult.value;
      }
      else {

        let internetValue = await internetPromise;
        idb.set(config.url, internetValue.value);
        return internetValue.value;
      }
    }
    else if (firstResult.type === 'internet') {
      console.log('Internet was faster than idb?')

      // Store value in idb, but don't wait for it to save before returning
      idb.set(config.url, firstResult.value);

      return firstResult.value;

    }
    else {
      console.error('ERROR?')
    }
  }
}

export default new Request();
