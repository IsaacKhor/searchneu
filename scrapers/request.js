/*
 * Copyright (c) 2017 Ryan Hughes
 *
 * This file is part of CoursePro.
 *
 * CoursePro is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License
 * version 3 as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import request from 'request-promise-native';
import URI from 'urijs';
import fs from 'fs-promise';
import async from 'async';
import dns from 'dns-then';
import mkdirp from 'mkdirp-promise';
import objectHash from 'object-hash';
import path from 'path';

import utils from './utils';
import macros from './macros';

// This file is a transparent wrapper around the request library that changes some default settings so scraping is a lot faster.
// This file adds:
// Automatic retry, with a delay in between reqeusts.
// Limit the max number of simultaneous requests (this used to be done with d3-queue, but is now done with agent.maxSockets)
// Application layer DNS cacheing. The request library (and the built in http library) will do a separate DNS lookup for each request
//    This DNS cacheing will do one DNS lookup per hostname (www.ccis.northeastern.edu, wl11gp.neu.edu) and cache the result.
//    This however, does change the URL that is sent to the request library (hostname -> pre-fetched ip), which means that the https verificaton will be comparing
//    the URL in the cert with the IP in the url, which will fail. Need to disable https verification for this to work.
// Keep-alive connections. Keep TCP connections open between requests. This significantly speeds up scraping speeds (1hr -> 20min)
// Ignore invalid HTTPS certificates and outdated ciphers. Some school sites have really old and outdated sites. We want to scrape them even if their https is misconfigured.

// In progress: save all pages to disk so development of parsers is faster and don't need to hit prod to test updates for scrapers

// This object must be created once per process
// Attributes are added to this object when it is used
// This is the total number of requests per host
// https://github.com/request/request
const separateReqDefaultPool = { maxSockets: 10000, keepAlive: true, maxFreeSockets: 10000 };

// Specific limits for some sites. CCIS will reject request if too many are made too quickly.
// Some other schools' servers will crash/slow to a crawl if too many requests are sent too quickly.
const separateReqPools = {
  'www.ccis.northeastern.edu': { maxSockets: 10, keepAlive: true, maxFreeSockets: 10 },
};

const MAX_RETRY_COUNT = 35;

// These numbers are in ms.
const RETRY_DELAY = 20000;
const RETRY_DELAY_DELTA = 15000;

class Request {

  constructor() {
    this.openRequests = 0;

    this.dnsPromises = {};
  }

  // By default, needle and nodejs does a DNS lookup for each request.
  // Avoid that by only doing a dns lookup once per domain
  async getDns(hostname) {
    if (this.dnsPromises[hostname]) {
      return this.dnsPromises[hostname];
    }

    // Just the host + subdomains are needed, eg blah.google.com
    if (hostname.startsWith('http://') || hostname.startsWith('https://')) {
      utils.elog(hostname);
    }

    const promise = dns.lookup(hostname, {
      all: true,
      family: 4,
    });

    this.dnsPromises[hostname] = promise;

    const result = await promise;

    if (result.length > 1) {
      console.log('INFO: more than 1 dns result', result);
    }

    return result;
  }


  standardizeInputConfig(config, method = 'GET') {
    if (typeof config === 'string') {
      config = {
        method: method,
        url: config,
      };
    }

    if (!config.headers) {
      config.headers = {};
    }

    return config;
  }


  async fireRequest(config) {
    config = this.standardizeInputConfig(config);


    // Default to JSON for POST bodies
    if (config.method === 'POST' && !config.headers['Content-Type']) {
      config.headers['Content-Type'] = 'application/json';
    }

    const urlParsed = new URI(config.url);

    const hostname = urlParsed.hostname();

    const dnsResults = await this.getDns(urlParsed.hostname());


    let ip;
    if (dnsResults.length === 0) {
      utils.elog('DNS lookup returned 0 results!', JSON.stringify(config));
      return null;
    } else if (dnsResults.length === 1) {
      ip = dnsResults[0].address;
    } else {
      const index = Math.floor(Math.random() * dnsResults.length);
      ip = dnsResults[index].address;
    }

    // Make the start of the new url with the ip from the DNS lookup and the protocol from the url
    const urlStart = new URI(ip).protocol(urlParsed.protocol()).toString();

    // Then add on everything after the host
    const urlWithIp = new URI(urlParsed.resource()).absoluteTo(urlStart).port(urlParsed.port()).toString();


    // Setup the default config
    // Change some settings from the default request settings for
    const defaultConfig = {
      headers: {},
    };

    // Default to JSON for POST bodies
    if (config.method === 'POST') {
      defaultConfig.headers['Content-Type'] = 'application/json';
    }

    // Enable keep-alive to make sequential requests faster
    if (separateReqPools[hostname]) {
      defaultConfig.pool = separateReqPools[hostname];
    } else {
      defaultConfig.pool = separateReqDefaultPool;
    }

    //Ten min
    defaultConfig.timeout = 60 * 10000;

    defaultConfig.resolveWithFullResponse = true;

    // Allow fallback to old depreciated insecure SSL ciphers. Some school websites are really old  :/
    // We don't really care abouzt security (hence the rejectUnauthorized:false), and will accept anything.
    // Additionally, this is needed when doing application layer dns caching because the url no longer matches the url in the cert
    defaultConfig.rejectUnauthorized = false;
    defaultConfig.ciphers = 'ALL';

    // Set the host in the header to the hostname on the url.
    // This is not done automatically because of the application layer dns caching (it would be set to the ip instead)
    defaultConfig.headers.Host = hostname;

    defaultConfig.headers['User-Agent'] = 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10.10; rv:24.0) Gecko/20100101 Firefox/24.0';


    //trololololol
    //Needed on some old sites that will redirect/block requests when this is not set
    //when a user is requesting a page that is not the entry page of the site
    //temple, etc
    defaultConfig.headers.Referer = config.url;


    // Merge the default config and the input config
    // Need to merge headers and output separately because config.headers object would totally override
    // defaultConfig.headers if merged as one object (Object.assign does shallow merge and not deep merge)
    const output = {};
    const headers = {};

    Object.assign(headers, defaultConfig.headers, config.headers);
    Object.assign(output, defaultConfig, config);

    output.url = urlWithIp;
    output.headers = headers;

    this.openRequests++;
    let response;
    try {
      response = await request(output);
    } catch (e) {
      this.openRequests --;
      throw e;
    }
    this.openRequests --;
    return response;
  }


  doAnyStringsInArray(array, body) {
    for (let i = 0; i < array.length; i++) {
      if (body.includes(array[i])) {
        return true;
      }
    }
    return false;
  }


  async request(config) {
    config = this.standardizeInputConfig(config);


    const urlParsed = new URI(config.url);

    let folder;
    let filename;
    let filePath;


    if (macros.DEV) {
      folder = path.join('request_cache', urlParsed.hostname());

      // Ensure only letters and numbers and dots and limit char length
      filename = urlParsed.path().replace(/[^A-Za-z0-9.]/gi, '_').trim().slice(0, 25) + objectHash(config);

      await mkdirp(folder);

      filePath = path.join(folder, filename);

      const exists = await fs.exists(filePath);

      if (exists) {
        const contents = JSON.parse((await fs.readFile(filePath)).toString());
        // console.log('Loaded ', contents.body.length, 'from cache', config.url);
        return contents;
      }
    }


    let tryCount = 0;

    return new Promise((resolve, reject) => {
      async.retry({
        times: MAX_RETRY_COUNT,
        interval: RETRY_DELAY + Math.round(Math.random() * RETRY_DELAY_DELTA),
      }, async (callback) => {
        let response;
        tryCount++;
        try {
          response = await this.fireRequest(config);
        } catch (err) {
          // Most sites just give a ECONNRESET or ETIMEDOUT, but dccc also gives a EPROTO and ECONNREFUSED.
          // This will retry for any error code.
          console.log('Try#:', tryCount, 'Code:', err.statusCode || err, ' Open request count: ', this.openRequests, 'Url:', config.url);
          callback(err);
          return;
        }

        // Ensure that body contains given string.
        if (config.requiredInBody && !this.doAnyStringsInArray(config.requiredInBody, response.body)) {
          console.log('Try#:', tryCount, 'Warning, body did not contain specified text', response.body.length, response.statusCode, this.openRequests, config.url);
          callback('Body missing required text.');
          return;
        } else if (response.body.length < 4000 && !config.shortBodyWarning === false) {
          console.log('Warning, short body', config.url, response.body, this.openRequests);
        }

        callback(null, response);
      }, async (err, response) => {
        if (err) {
          reject(err);
          return;
        }

        // Save the response to a file for development
        if (macros.DEV) {
          await fs.writeFile(filePath, JSON.stringify(response.toJSON()));
        }

        console.log('Parsed', response.body.length, 'from ', config.url);
        resolve(response);
      });
    });
  }

  // Helpers for get and post
  async get(config) {
    if (!config) {
      console.log('Warning, request called with no config');
      return null;
    }
    if (typeof config === 'string') {
      return this.request({
        url: config,
        method: 'GET',
      });
    }

    config.method = 'GET';
    return this.request(config);
  }

  async post(config) {
    if (!config) {
      console.log('Warning, request called with no config');
      return null;
    }
    if (typeof config === 'string') {
      return this.request({
        url: config,
        method: 'POST',
      });
    }

    config.method = 'POST';
    return this.request(config);
  }

  async head(config) {
    if (!config) {
      console.log('Warning, request called with no config');
      return null;
    }
    if (typeof config === 'string') {
      return this.request({
        url: config,
        method: 'HEAD',
      });
    }

    config.method = 'HEAD';
    return this.request(config);
  }

  // Do a head request. If that fails, do a get request. If that fails, the site is down and return false
  // need to turn off high retry count
  async isPageUp(config) {
    throw new Error('This does not work yet');
    // this.head(config)
  }


}


// async function test() {
//   const it = new Request();

//   try {
//     const d = await it.request({
//       url:'http://localhost',
//     });

//     console.log(d.body, 'NO CRASH');
//   } catch (e) {
//     console.log(e, 'HERE');
//   }
// }


const instance = new Request();



export default instance;
