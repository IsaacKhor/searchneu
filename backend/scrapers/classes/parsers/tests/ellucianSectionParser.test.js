/*
 * Copyright (c) 2017 Ryan Hughes
 *
 * This file is part of CoursePro.
 *
 * CoursePro is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License
 * version 3 as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>. 
 */

import path from 'path';
import fs from 'fs-promise';

import request from '../../../request';

var ellucianSectionParser = require('../ellucianSectionParser')


it('parse prereqs and coreqs and seat data from 1.html', async function (done) {

  //the pre and co requs html here has been modified
  //this contains the same pre requs as prereqs10
  const body = await fs.readFile(path.join(__dirname, 'data', 'ellucianSectionParser', '1.html'), 'utf8');

  var url = 'https://wl11gp.neu.edu/udcprod8/bwckschd.p_disp_detail_sched?term_in=201610&crn_in=15633';

  expect(ellucianSectionParser.supportsPage(url)).toBe(true);

  const retVal = ellucianSectionParser.parse(body, url);

  expect(retVal).toMatchSnapshot()

  done()
});




it('honors works', async function (done) {

  const body = await fs.readFile(path.join(__dirname, 'data', 'ellucianSectionParser', 'honors.html'), 'utf8')

  var url = 'https://wl11gp.neu.edu/udcprod8/bwckschd.p_disp_detail_sched?term_in=201610&crn_in=15633';

  expect(ellucianSectionParser.supportsPage(url)).toBe(true);

  let retVal = ellucianSectionParser.parse(body, url);

  expect(retVal).toMatchSnapshot();

  done()
});
