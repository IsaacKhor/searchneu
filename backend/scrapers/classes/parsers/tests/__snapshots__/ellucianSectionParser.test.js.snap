// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`honors works 1`] = `
Object {
  "honors": true,
  "isOnline": false,
  "maxCredits": 0,
  "minCredits": 0,
  "seatsCapacity": 32,
  "seatsRemaining": 31,
  "waitCapacity": 0,
  "waitRemaining": 0,
}
`;

exports[`parse prereqs and coreqs and seat data from 1.html 1`] = `
Object {
  "coreqs": Object {
    "type": "or",
    "values": Array [
      Object {
        "classId": "2161",
        "subject": "EECE",
      },
    ],
  },
  "honors": true,
  "isOnline": false,
  "maxCredits": 3,
  "minCredits": 3,
  "prereqs": Object {
    "type": "and",
    "values": Array [
      Object {
        "type": "or",
        "values": Array [
          Object {
            "classId": "1601",
            "subject": "AE",
          },
          Object {
            "classId": "1350",
            "subject": "AE",
          },
        ],
      },
      Object {
        "type": "or",
        "values": Array [
          Object {
            "classId": "2212",
            "subject": "PHYS",
          },
          Object {
            "classId": "2232",
            "subject": "PHYS",
          },
        ],
      },
      Object {
        "type": "or",
        "values": Array [
          Object {
            "classId": "2401",
            "subject": "MATH",
          },
          Object {
            "classId": "2411",
            "subject": "MATH",
          },
          Object {
            "classId": "24X1",
            "subject": "MATH",
          },
          Object {
            "classId": "2551",
            "subject": "MATH",
          },
          Object {
            "classId": "2561",
            "subject": "MATH",
          },
          Object {
            "classId": "2X51",
            "subject": "MATH",
          },
        ],
      },
      Object {
        "classId": "2001",
        "subject": "COE",
      },
    ],
  },
  "seatsCapacity": 32,
  "seatsRemaining": 0,
  "waitCapacity": 0,
  "waitRemaining": 0,
}
`;
